include/BingoCpp/gradient_mixin.h

```diff
@@ -0,0 +1,61 @@
+#ifndef BINGOCPP_INCLUDE_BINGOCPP_GRADIENT_MIXIN_H_
+#define BINGOCPP_INCLUDE_BINGOCPP_GRADIENT_MIXIN_H_
+
+#include <Eigen/Dense>
+#include <bingocpp/equation.h>
+#include <bingocpp/fitness_function.h>
+
+#include <functional>
+#include <cmath>
+
+namespace bingo {
+
+class GradientMixin {
+ public:
+  virtual std::tuple<double, Eigen::VectorXd> GetIndividualFitnessAndGradient(Equation &individual) const = 0;
+};
+
+class VectorGradientMixin : public GradientMixin {
+ public:
+  VectorGradientMixin(TrainingData *training_data = nullptr, std::string metric = "mae");
+
+  std::tuple<double, Eigen::VectorXd> GetIndividualFitnessAndGradient(Equation &individual) const;
+
+  virtual std::tuple<Eigen::VectorXd, Eigen::ArrayXXd> GetFitnessVectorAndJacobian(Equation &individual) const = 0;
+
+ protected:
+  static double mean_absolute_error(const Eigen::VectorXd &fitness_vector) {
+    return fitness_vector.array().abs().mean();
```

Comment:
This line got me thinking, is a VectorXd really the right choice of type here?  are we ever taking advantage of the vector operations in Eigen? or are we always doing array-wise operations?

From what I can tell in the eigen docs, there shouldn't be any runtime cost of using a vector instead, but it doesn't seem as clean.  Changing to an ArrayXd is probably my preference, but this might be somewhat of a large overhaul as we would probably want to change throughout the codebase.  The ArrayXd might also potentially set us up for future improvements in the vectorization of the evaluation backend.

if we don't want to add this change to the current merge lets make an issue and circle back at a later point

Reply:
Yeah I like using ArrayXd better here also. I was trying to stay consistent with the usage of VectorXd in [EvaluateFitnessVector](https://github.com/nasa/bingocpp/blob/develop/include/bingocpp/fitness_function.h#L88). However, I can refactor to using ArrayXd instead.

As you mentioned, the change will probably span multiple files. I can change to using ArrayXd in the current changes then add the refactored changes to another pull request if that would be easier for you to review. If not, I will just add the changes onto this pull request. Let me know what you prefer!